### DISCUSSION TITLE
about --keep N

### DISCUSSION THREAD

**Discussion by maddes8cht** [ggml-org/llama.cpp#2289] [Q&A]
I'm still confused about the `--keep` flag, as i mentioned in #46.
Always when `--keep N` is mentioned, it is stated 
> number of tokens to keep from the initial prompt (default: 0, -1 = all)

To me, this sounds exactly like what I would expect from a "system" message:
The `initial` prompt, the very first prompt of the conversation, should always remain in context. This is how I understand this line.
In my attempts I always had the impression that this does not work and that the `initial` prompt always disappears from the context.
For clarification:
Is it rather the case that `--keep` merely controls that in a concrete answer the user's previously requested prompt remains in the context (to the selected portion)?  (And in contrast to the ìnitial`, i.e. the very first prompt?)
While older parts of the context are nevertheless pushed out?
Such a setting makes sense, too, of course - also and especially together with a system prompt that does not disappear from the context even in a longer conversation.

**Comment by halfburnttoast** [comment]
I’d like to know too. If it is supposed to keep a static context throughout the entire session, it’s definitely not working. I was under the impression that will keep the initial prompt in context at all times. But, during my testing, it definitely drops it.

**Comment by klosax** [comment]
Here is how it is supposed to work:
https://github.com/ggerganov/llama.cpp/blob/b9b7d94fc10a8039befd1bc3af4f4b09c620c351/examples/main/main.cpp#L402-L405

The initial/original prompt is supplied with the -p parameter when you start the session.

**Comment by maddes8cht** [comment]
so there still remains the question why it doesn't seem to work?
